### ISSUE TITLE
Make optional() properties optional in typescript

### ISSUE THREAD

**Issue by calummoore** [neuledge/computed-types#51]
Currently, when I make a property optional in the schema it outputs the type as undefined - which is not quite the same as optional.

```tsx
export const PersonSchema = Schema({
  name: string.optional(),
})

export type SourceMetric = Type<typeof PersonSchema>

const obj: MySchema {} // this would error, missing name

const obj: MySchema {
  name: undefined // allowed
}
```

As per this comment, it's fairly easy to add a utility type function that can convert undefined to optional. I think it would make sense to apply this by default as part of `Type`. 

https://github.com/Microsoft/TypeScript/issues/12400#issuecomment-502345531

**Comment by moshest** [comment]
Hmm.. Computed types is based on function interfaces. Specifically, it will generate the type from the function return type and generate an input type from the function first argument.

I think the input types are matching exactly the function input, so optional arguments will convert to optional inputs:

```ts
const Demo = Schema({
  foo: (input?: 'bar'): true | undefined => (!!input || undefined),
});

Parameters<typeof Demo> === [{ foo?: 'bar' }]
Type<typeof Demo> === { foo: true | undefined }
```

It's true that currently the handling of the return type is not handled optional values, as functions always returns a value.
I think the best solution for this is to check if `(validator argument is optional) && (validator can return undefined)`.

It wont solve all the cases, but it will solve most cases.

Thoughts?

**Comment by calummoore** [comment]
Yes - I think that would work. 👍

In reality, there is no difference between undefined and optional. I think typescript was planning to make them the same, but just haven't got around to implementing it yet.

**Comment by SirJosh3917** [comment]
Any update on this issue? My usecase is that I currently have a schema that looks like this
```ts
export const Block = Schema({
    id: 1
});
```

I want to update this schema in a backwards compatible way.

```ts
export const Block = Schema({
    id: 1,
    color: Schema.either("white" as const, "black" as const).optional()
});
```

such that no code is broken. However, due to `optional` just appending `| undefined`, it requires clients to specify `color`. Currently my workaround is to manually play with the type so that the optional is in the final value, but this is unmaintainable and defeats the purpose of using this library.

**Comment by moshest** [comment]
It's looks like that what you need here is to use the `SchemaInput<S>` instead of `Type<S>` (available since v1.2)

`SchemaInput ` will return you the actual type the schema willing to accept, in this case it will be:

```ts
type {
  id: number,
 color?: 'white' | 'black',
}
```

Changing the Type and omitting undefined values is not trivial will probably we on the next breaking version and I hope it will be within a week or two.
